docs/rfcs/014-storage-lsm.md

```diff
@@ -0,0 +1,146 @@
+# Why LSM trees?
+
+In general, an LSM tree has the nice property that random updates are
+fast, but the disk writes are sequential. When a new file is created,
+it is immutable. New files are created and old ones are deleted, but
+existing files are never modified. That fits well with storing the
+files on S3.
+
+Currently, we create a lot of small files. That is mostly a problem
+with S3, because each GET/PUT operation is expensive. Currently, the
```

Comment:
@antons-antons can you paste a link here if you find the cost of multipart uploads? Not sure if a paginated upload counts as one PUT. Maybe we should try and see how much we get billed :)

Reply:
I agree that language in the RFC seems a little vague regarding the "**_small_**" and "**_expensive_**", with pricing since Dec 2021:
Storage $0.01 between $0.023 per GB (depending on scale, access pattern, region)
* **PUT**, COPY, POST, **LIST**: $0.005-$0.01 per 1K requests
* **GET**, SELECT: $0.0004-$0.001 per 1K requests
* Delete - free

with storage instances (assuming 3yr reserved capacity, 30% discount vs on demand):
* is4gen.8XL: 30000GB, $2264 per month -> $0.075 per GB
* im4ng.16XL: 30000GB, $2687 per month -> $0.09 per GB
* i3en.metal: 60000GB, $5319 per month -> $0.088 per GB

It looks to me that aggressive offload (more granular?) with page faulting is likely a more economic solution.


Regarding specific details on API cost for multi-part upload (e.g. when AWS cli  decides to issue a multi-part upload) AWS documentation is vague.
But rust-S3 implementation it seems that multi-part upload consists of Post for InitiateMultipartUpload + Post for CompleteMultipartUpload + 1 PUT per object chunk (keep in mind retries will incur extra cost).
Current implementation in rust-S3::Bucket::put_object uses single chunk uploads.

Additionally it's possible to implement a parallel download with each request getting billed separately.

meanwhile multi-part uploads could give us parallelism (and a better error granularity) at a cost of complexity and extra cleanup:
https://aws.amazon.com/blogs/aws-cloud-financial-management/discovering-and-deleting-incomplete-multipart-uploads-to-lower-amazon-s3-costs/

Reply:
`ListObjectsV2` has a limit of 1000 objects returned per request; so any actual *LIST* operation will cost at least 0.005$ per million objects listed (2 times as expensive as S3 Inventory; which operates at the bucket level but with fewer details returned).
This is not a lot _per se_; but it will become the dominant portion of S3 costs if each month we list a file many times. (the cutoff depends on the storage method and the region; but for Standard this fraction is 1/4.6; i.e. 10k times a full listing of a bucket each month will be nearly as expensive as the storage costs of that bucket if the average size of files in that bucket is 10k * 1kB / 4.6 ~= 2MB.

For a full LIST operation each minute (at 43800 minutes per standard-month) LIST operations become more expensive than storage costs when the average size of objects is below 9.5MB (Standard), 17.5MB (Infrequent Access), or 54MB (Glacier Instant Retrieval). ("file size with storage cost == list cost" formula is `<n LISTs /month> * 0.005$ /<storage class cost in $/GB/month> * 1kB`)

Note that even before object size is below the given size that LIST costs are quite significant; just not as significant as the S3 storage costs.

Reply:
> It looks to me that aggressive offload (more granular?) with page faulting is likely a more economic solution.
But can we provide reasonable response time with such solution?
Please notice that we are talking about OLTP, not OLAP. Even if database is rarely used (few requests per day), 
still few seconds delay for query may me unacceptable: in most client application such database request will be interrupted by timeout. And users also will not happy if response for their actions is not received in less than one second.